File: erlang/hamming/src/hamming.erl
-module(hamming).

-export([distance/2, test_version/0]).


distance(Strand1, Strand2) -> distance(Strand1, Strand2, 0).

distance([],[],Distance) -> Distance;
distance(_,[],_) -> {error, "left and right strands must be of equal length"};
distance([],_,_) -> {error, "left and right strands must be of equal length"};
distance([H|T1],[H|T2],Distance) -> distance(T1, T2, Distance); 
distance([H1|T1],[H2|T2],Distance) -> distance(T1, T2, Distance + 1).


test_version() -> 2.

File: go/hamming/hamming.go
// Package calculates the hamming hamlength between two DNA strings
package hamming

import "errors"

// Distance between two DNA strings, as an integer.
func Distance(a, b string) (int, error) {
    
    var hamlength int
    var err error

    // Trivial statement: unequal strand length
	if len(a) != len(b) {
		return -1, errors.New("strand lengths must be identical")
	}

	// Equal strand length
    for i := range a {
		if a[i] != b[i] {
			hamlength++
		}
	}

	return hamlength, err


}
